diff --git a/meson.build b/meson.build
index d45ac05..4d619e9 100644
--- a/meson.build
+++ b/meson.build
@@ -258,11 +258,4 @@ config_h_include = include_directories('.')
 
 # Subdirs
 subdir('src')
-subdir('common')
-subdir('tools')
-if not get_option('test').disabled()
-  subdir('test')
-endif
-if doxygen.found()
-  subdir('doc')
-endif
+subdir('common')
\ No newline at end of file
diff --git a/src/openslide-decode-tifflike.c b/src/openslide-decode-tifflike.c
index fb254c8..1c3c976 100644
--- a/src/openslide-decode-tifflike.c
+++ b/src/openslide-decode-tifflike.c
@@ -698,8 +698,6 @@ static void print_tag(struct _openslide_tifflike *tl,
   struct tiff_item *item = get_item(tl, dir, tag);
   g_assert(item != NULL);
 
-  printf(" %d: type: %d, count: %"PRId64"\n ", tag, item->type, item->count);
-
   switch (item->type) {
   case TIFF_ASCII: {
     // will only print first string if there are multiple
